ts/src/base/Exchange.ts

```diff
@@ -3963,38 +3962,11 @@ export default class Exchange {
         if (vwap === undefined) {
             vwap = Precise.stringDiv (this.omitZero (quoteVolume), baseVolume);
         }
-        if ((last !== undefined) && (close === undefined)) {
-            close = last;
-        } else if ((last === undefined) && (close !== undefined)) {
-            last = close;
-        }
-        if ((last !== undefined) && (open !== undefined)) {
-            if (change === undefined) {
-                change = Precise.stringSub (last, open);
-            }
-            if (average === undefined) {
-                let precision = 18;
-                if (market !== undefined && this.isTickPrecision ()) {
-                    const marketPrecision = this.safeDict (market, 'precision');
-                    const precisionPrice = this.safeString (marketPrecision, 'price');
-                    if (precisionPrice !== undefined) {
-                        precision = this.precisionFromString (precisionPrice);
-                    }
-                }
-                average = Precise.stringDiv (Precise.stringAdd (last, open), '2', precision);
-            }
-        }
-        if ((percentage === undefined) && (change !== undefined) && (open !== undefined) && Precise.stringGt (open, '0')) {
-            percentage = Precise.stringMul (Precise.stringDiv (change, open), '100');
-        }
-        if ((change === undefined) && (percentage !== undefined) && (open !== undefined)) {
-            change = Precise.stringDiv (Precise.stringMul (percentage, open), '100');
-        }
-        if ((open === undefined) && (last !== undefined) && (change !== undefined)) {
-            open = Precise.stringSub (last, change);
-        }
+        [ close, open, change, percentage, average ] = this.safeTickerCalculationHelper (open, close, change, percentage, average, market);
+        [ close, open, change, percentage, average ] = this.safeTickerCalculationHelper (open, close, change, percentage, average, market);
```

Comment:
we need to call `safeTickerCalculationHelper` twice, because some fields are calculated in the bottom of "auto-calculations" however, those fields are needed to derive other values. so, only one cycle is not enough.